### PR TITLE
Dp 408/set headers to requests

### PR DESCRIPTION
**Pull request by ravitb**
#### Documentation (Source/readme.md)

### CODE REVIEW

**Comment by ravitb** [comment]
@yosriz, please notice that the tests are failing because of the dependency in js-node-sdk which is also needs a review.

File: tests/src/kinAccount.intg.test.ts [hunk 1]
```diff
@@ -171,5 +170,14 @@ describe("KinAccount", async () => {
...
 	}, 60000);
 
+	test("Create account with an existing address, expect error", async () => {
+
+		const txBuilder = await sender.buildCreateAccount({
+			address: receiver.publicAddress,
+			startingBalance: 10,
+			fee: 100
+		});
+		expect((txBuilder._server as any).headers.get("USER-AGENT")).toEqual(Server.GLOBAL_HEADERS.get("USER-AGENT"));
```
Comment (yosriz) [inline tests/src/kinAccount.intg.test.ts]: not really testing the behavior (just the assignment), also integ test is not needed for that anyway, I would do a unit test with nock, telling nock to expect to get the input headers when reaching the network.

File: package.json [hunk 2]
```diff
@@ -4,6 +4,9 @@
   "description": "Kin SDK for the Node runtime",
   "main": "scripts/bin/index.js",
   "types": "scripts/bin/index.d.ts",
+  "runtimeArgs": [
```
Comment (yosriz) [inline package.json]: should be removed (local change for npm link?)

File: scripts/src/index.ts [hunk 3]
```diff
@@ -40,8 +40,10 @@ import {
 	ResourceNotFoundError,
 	ServerError
 } from "./errors";
+import {Server} from "./server";
 
 export {
+	Server,
```
Comment (yosriz) [inline scripts/src/index.ts]: any reason to export `Server` as an API?

File: scripts/src/server.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+const PACKAGE_JSON = require('../../package.json');
+import {Server as KinServer} from "@kinecosystem/kin-sdk";
+
+export class Server extends KinServer {
```
Comment (yosriz) [inline scripts/src/server.ts]: why we need this inheritance?
There's one place (`KinClient`) that instantiate `Server` where we need to add the headers and that's all.
it makes a confuse regarding the name (Server vs KinServer)

File: tests/src/channels.unit.test.ts [hunk 5]
```diff
@@ -25,7 +25,7 @@ describe("Channels", async () => {
 
 	beforeAll(async () => {
 		const keyPair = KeyPair.fromSeed(senderSeed);
-		const server = new Server(fakeUrl, {allowHttp: true});
+		const server = new Server(fakeUrl, {allowHttp: true, headers: Server.GLOBAL_HEADERS});
```
Comment (yosriz) [inline tests/src/channels.unit.test.ts]: do we need to add the headers in all tests now? seems it can be left intact.

File: package.json (hunk 2 above)
Comment (ravitb) [reply package.json]: Exactly.

File: scripts/src/blockchain/transactionBuilder.ts [hunk 6]
```diff
@@ -14,7 +16,8 @@ export class TransactionBuilder {
 	private readonly _channel?: Channel;
 	private readonly _appId?: string;
 
-	constructor(readonly _server: Server, sourceAccount: Account, options: TransactionBuilderOptions, channel?: Channel) {
+	constructor(sourceAccount: Account, options: TransactionBuilderOptions, channel?: Channel) {
+		options.headers = options.headers ? options.headers : GLOBAL_HEADERS;
```
Comment (yosriz) [inline scripts/src/blockchain/transactionBuilder.ts]: headers are not needed in this class, transaction builder doesn't use headers

File: scripts/src/kinClient.ts [hunk 7]
```diff
@@ -27,9 +27,9 @@ export class KinClient {
 	private readonly _transactionRetriever: TransactionRetriever;
 	private readonly _blockchainListener: BlockchainListener;
 
-	constructor(private readonly _environment: Environment) {
+	constructor(private readonly _environment: Environment, opts = {allowHttp: false, headers: GLOBAL_HEADERS}) {
```
Comment (yosriz) [inline scripts/src/kinClient.ts]: opts shouldn't be a new param to `KinClient`, this is an internal detail

File: tests/src/accountDataRetriever.unit.test.ts [hunk 8]
```diff
@@ -25,7 +26,7 @@ const mock500NetworkResponse: ErrorResponse = {
 
 describe("AccountDataRetreiver.fetchAccountData", async () => {
 	beforeAll(async () => {
-		accountDataRetriever = new AccountDataRetriever(new Server(fakeUrl, {allowHttp: true}));
+		accountDataRetriever = new AccountDataRetriever(new Server(fakeUrl, {allowHttp: true, headers: new Map<string, string>().set("user-agent", "test-kin")}));
```
Comment (yosriz) [inline tests/src/accountDataRetriever.unit.test.ts]: headers are not required for the tests in this file

File: tests/src/channels.unit.test.ts [hunk 9]
```diff
@@ -7,10 +7,10 @@ import {ChannelsPool} from "../../scripts/src/blockchain/channelsPool";
 import * as nock from "nock";
 import {TransactionBuilder} from "../../scripts/src/blockchain/transactionBuilder";
 import {Environment} from "../../scripts/src/environment";
+import {GLOBAL_HEADERS} from "../../scripts/src/config";
```
Comment (yosriz) [inline tests/src/channels.unit.test.ts]: import can be removed

File: tests/src/channels.unit.test.ts [hunk 10]
```diff
@@ -25,7 +25,7 @@ describe("Channels", async () => {
 
 	beforeAll(async () => {
 		const keyPair = KeyPair.fromSeed(senderSeed);
-		const server = new Server(fakeUrl, {allowHttp: true});
+		const server = new Server(fakeUrl, {allowHttp: true, headers: new Map<string, string>().set("user-agent", "test-kin")});
```
Comment (yosriz) [inline tests/src/channels.unit.test.ts]: headers not needed here as well

File: tests/src/kinAccount.intg.test.ts [hunk 11]
```diff
@@ -3,6 +3,7 @@ import {Environment} from "../../scripts/src/environment";
 import {Channels, KeyPair, KinClient} from "../../scripts/src";
 import {Keypair, Memo, Network, Operation, Transaction as XdrTransaction} from "@kinecosystem/kin-base";
 import {Server} from "@kinecosystem/kin-sdk";
+import {GLOBAL_HEADERS} from "../../scripts/src/config";
```
Comment (yosriz) [inline tests/src/kinAccount.intg.test.ts]: remove

File: tests/src/kinClient.intg.test.ts [hunk 12]
```diff
@@ -131,13 +131,13 @@ describe("KinClient", async () => {
...
 		let sendBuilder = await sender.buildSendKin({
 			address: receiver.publicAddress,
 			amount: 10,
 			fee: 100,
 			memoText: 'sending kin'
 		});
 		hash = await sender.submitTransaction(sendBuilder);
-	}, 120000);
+
+	}, 180000);
```
Comment (yosriz) [inline tests/src/kinClient.intg.test.ts]: why increasing so much the timeout?

File: tests/src/transactionRetrieverTest_fetchTransactionHistory.unit.test.ts [hunk 13]
```diff
@@ -1,9 +1,10 @@
-import {Server} from "@kinecosystem/kin-sdk";
 import {TransactionRetriever} from "../../scripts/src/blockchain/transactionRetriever";
 import * as nock from "nock";
 import {CreateAccountTransaction, PaymentTransaction, RawTransaction} from "../../scripts/src/blockchain/horizonModels";
 import {ErrorResponse, InternalError, ResourceNotFoundError} from "../../scripts/src/errors";
 import {Memo, Operation} from "@kinecosystem/kin-base";
+import {Server} from "@kinecosystem/kin-sdk";
+import {GLOBAL_HEADERS} from "../../scripts/src/config";
```
Comment (yosriz) [inline tests/src/transactionRetrieverTest_fetchTransactionHistory.unit.test.ts]: remove (optimize imports before a commit can prevent those leftovers)

File: tests/src/kinAccount.unit.test.ts [hunk 14]
```diff
@@ -243,18 +245,21 @@ describe("KinAccount.createAccount", async () => {
 
 function mock400SendKinResponse(response: ErrorResponse) {
 	nock(fakeUrl)
+		.matchHeader(headerKey, headerVal)
```
Comment (yosriz) [inline tests/src/kinAccount.unit.test.ts]: 👍 now we're testing the headers.
In unit test, I would do a dedicate test just for headers, this is helpful for both isolation (adds noise to other tests) and locality (pinpoint the issue)

File: tests/src/kinClient.intg.test.ts (hunk 12 above)
Comment (ravitb) [reply tests/src/kinClient.intg.test.ts]: Sometimes it fails if the timeout is smaller.

**Pull request merged**